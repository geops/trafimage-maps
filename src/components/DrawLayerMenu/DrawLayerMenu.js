import React, { useMemo } from 'react';
import { useSelector } from 'react-redux';
import { useTranslation } from 'react-i18next';
import LayerTree from 'react-spatial/components/LayerTree';
import InfosButton from '../InfosButton';

function DrawLayerMenu() {
  const { t } = useTranslation();
  const layers = useSelector((state) => state.map.layers);
  const drawLayer = useSelector((state) => state.map.drawLayer);
  const drawIds = useSelector((state) => state.app.drawIds);

  const titles = useMemo(() => {
    return {
      layerShow: t('Layer anzeigen'),
      layerHide: t('Layer verbergen'),
    };
  }, [t]);

  if (!drawIds) {
    return null;
  }

  return (
    <div className="wkp-draw-layer-menu">
      <div className="wkp-layer-tree">
        <LayerTree
          isItemHidden={(layer) => layer !== drawLayer}
          renderItemContent={(layer, layerTreeComp) => (
            <>
              {layer.renderItemContent
                ? layer.renderItemContent(layerTreeComp)
                : layerTreeComp.renderItemContent(layer)}
              {layer.get('hasInfos') && <InfosButton selectedInfo={layer} />}
            </>
          )}
          layers={layers}
          titles={titles}
          t={t}
        />
      </div>
    </div>
  );
}

export default React.memo(DrawLayerMenu);
