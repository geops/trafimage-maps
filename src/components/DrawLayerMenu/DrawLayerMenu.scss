.wkp-draw-layer-menu {
  .wkp-layer-tree,
  .wkp-layer-tree .rs-layer-tree-item {
    margin: 0;
  }

  .rs-layer-tree .rs-layer-tree-item > :last-child {
    margin-right: 0;
  }

  .wkp-share-permalink-button .tm-button {
    width: auto;
  }

  /* TODO: THis css should not exist, fix the css of layer tree in react-satial */
  .rs-layer-tree .rs-layer-tree-item div.wkp-tooltip {
    position: fixed;

    .wkp-arrow-left,
    .wkp-close-bt {
      position: absolute;
    }
  }
}
