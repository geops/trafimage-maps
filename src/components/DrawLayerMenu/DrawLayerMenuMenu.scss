.wkp-draw-layer-menu {
  .wkp-layer-tree {
    margin: 0;
  }
}
