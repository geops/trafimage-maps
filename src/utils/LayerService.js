/**
 * A layer service class to handle layer adding, removing and visiblity.
 */
export default class LayerService {
  constructor(layers) {
    this.layers = layers;
  }

  getLayers() {
    return this.layers;
  }

  setLayers(layers) {
    this.layers = layers;
  }

  getLayersAsFlatArray(optLayers) {
    let layers = [];
    (optLayers || this.getLayers() || []).forEach((l) => {
      layers.push(l);
      const { children } = l;
      layers = layers.concat(this.getLayersAsFlatArray(children || []));
    });
    return layers;
  }

  getLayer(name) {
    return this.getLayersAsFlatArray().find((l) => {
      return l.name === name;
    });
  }

  getParent(child) {
    return this.getLayersAsFlatArray().find((l) => {
      return !!l.children.includes(child);
    });
  }

  getParents(child) {
    let layer = child;
    const parents = [];

    let parentLayer;
    do {
      parentLayer = this.getParent(layer);
      if (parentLayer) {
        parents.push(parentLayer);
        layer = parentLayer;
      }
    } while (parentLayer);

    return parents;
  }

  getBaseLayers() {
    return this.getLayersAsFlatArray().filter((l) => {
      return l.get('isBaseLayer');
    });
  }

  getQueryableLayers() {
    return this.getLayersAsFlatArray().filter((layer) => {
      return layer.visible && layer.get('isQueryable');
    });
  }

  getFeatureInfoAtCoordinate(coordinate, layers) {
    const promises = (layers || this.getQueryableLayers()).map((layer) => {
      return layer
        .getFeatureInfoAtCoordinate(coordinate)
        .then((featureInfo) => {
          return featureInfo;
        });
    });
    return Promise.all(promises);
  }
}
