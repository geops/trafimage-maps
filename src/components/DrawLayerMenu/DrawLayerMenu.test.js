import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import configureStore from 'redux-mock-store';
import renderer from 'react-test-renderer';
import LayerService from 'react-spatial/LayerService';
import { mount } from 'enzyme';
import { Layer } from 'mobility-toolbox-js/ol';
import OLLayer from 'ol/layer/Layer';
import DrawLayerMenu from './DrawLayerMenu';

describe('DrawLAyerMenu', () => {
  const mockStore = configureStore([thunk]);
  let store;

  beforeEach(() => {
    store = mockStore({
      map: {},
      app: {},
    });
  });

  describe('should match snapshot.', () => {
    test('using the layerService property', () => {
      const layerService = new LayerService([]);
      const component = renderer.create(
        <Provider store={store}>
          <DrawLayerMenu layerService={layerService} />
        </Provider>,
      );
      const tree = component.toJSON();
      expect(tree).toMatchSnapshot();
    });
  });

  test('add a layer to the layerService', () => {
    const layerService = new LayerService([]);
    mount(
      <Provider store={store}>
        <DrawLayerMenu layerService={layerService} />
      </Provider>,
    );
    expect(layerService.layers.length).toBe(1);
    expect(layerService.layers.pop().get('hideInLegend')).toBe(true);
  });

  test('display only draw layer', () => {
    const layerService = new LayerService([
      new Layer({ olLayer: new OLLayer({}) }),
    ]);
    const wrapper = mount(
      <Provider store={store}>
        <DrawLayerMenu layerService={layerService} />
      </Provider>,
    );
    expect(layerService.layers.length).toBe(2);
    expect(wrapper.find('.rs-layer-tree-item').length).toBe(1);
  });

  test('hide draw layer in topic menus ', () => {
    const layerService = new LayerService([]);
    mount(
      <Provider store={store}>
        <DrawLayerMenu layerService={layerService} />
      </Provider>,
    );
    expect(layerService.layers.pop().get('hideInLegend')).toBe(true);
  });
});
