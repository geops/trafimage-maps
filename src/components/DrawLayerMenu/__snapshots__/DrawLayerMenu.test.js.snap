// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawLayerMenu should match snapshot. should return null 1`] = `""`;

exports[`DrawLayerMenu should match snapshot. using the layerService property 1`] = `
<div class="wkp-draw-layer-menu">
  <div class="wkp-layer-tree">
    <div class="rs-layer-tree">
    </div>
  </div>
</div>
`;
